agent/drive_handler.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/firecracker-microvm/firecracker-containerd/internal"
+)
+
+const (
+	blockPath = "/sys/block"
+
+	blockMajorMinor = "dev"
+	blockSize       = "size"
+)
+
+type drive struct {
+	Name       string
+	DriveID    string
+	MajorMinor string
+	Size       int
+	Type       string
+}
+
+type driveHandler struct {
+	drives map[string]drive
+}
+
+// DiscoverDrives will iterate the block path in the sys directory to retrieve all
+// stub block devices.
+func (d *driveHandler) DiscoverDrives() error {
+	names, err := getListOfBlockDeviceNames()
+	if err != nil {
+		return err
+	}
+
+	drives := map[string]drive{}
+	for _, name := range names {
+		d, err := buildDrive(name)
+		if err != nil {
+			return err
+		}
+
+		drives[name] = d
+	}
+
+	for k, v := range drives {
+		if !isStubDrive(v) {
```

Comment:
I can see why `isStubDrive` just returns false if an error occurs during its check, but I think we need to at least log those occurrences even if we still just return false.

Right now, if there was a bug that caused reading all drives to fail, we'd just silently hit this codepath for every drive and move on like nothing happened, which would probably result in confusing errors downstream of the actual problem here.